### PR TITLE
Water Supply/Contract DAO

### PR DESCRIPTION
**Pull request by zack-rma**
Created DAO for water supplies to manage water contracts and related information

### CODE REVIEW

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+import cwms.cda.data.dto.CwmsId;
+import cwms.cda.data.dto.LookupType;
+import cwms.cda.data.dto.watersupply.PumpType;
+import cwms.cda.data.dto.watersupply.WaterUser;
+import cwms.cda.data.dto.watersupply.WaterUserContract;
+import java.util.ArrayList;
+import java.util.List;
+import org.jooq.DSLContext;
+import org.jooq.impl.DSL;
+import usace.cwms.db.dao.ifc.loc.LocationRefType;
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: `usace.cwms.db.dao.*` and `usace.cwms.db.jooq.dao` should not be used.

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+        LocationRefType locationRefType = WaterSupplyUtils.map(projectLocation);
+
+        waterUserContractTypes.addAll(connectionResult(dsl, c -> {
+            setOffice(c, projectLocation.getOfficeId());
+            LOCATION_REF_T projectLocationRef =  LocationTypeUtil.toLocationRefT(locationRefType);
+            WATER_USER_CONTRACT_TAB_T waterUserContractObjTs = CWMS_WATER_SUPPLY_PACKAGE.call_RETRIEVE_CONTRACTS(
+                    DSL.using(c).configuration(), projectLocationRef, entityName);
+            return waterUserContractObjTs.stream()
+                    .map(WaterUserTypeUtil::toWaterUserContractType)
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: You're double mapping, going from the jOOQ codegen type to the cwms-db-jooq library (which we don't want to use) and then to CDA data type in the for loop below.

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                    .collect(toList());
+        }));
+
+        for (WaterUserContractType waterUserContractType : waterUserContractTypes) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: The mapping can be handled within the connectionResult block

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        List<usace.cwms.db.dao.ifc.cat.LookupType> contractTypeList = connectionResult(dsl, c -> {
+            setOffice(c, officeId);
+            LOOKUP_TYPE_TAB_T lookupTypeObjTs = CWMS_WATER_SUPPLY_PACKAGE.call_GET_CONTRACT_TYPES(
+                    DSL.using(c).configuration(), officeId);
+            return lookupTypeObjTs.stream()
+                    .map(LookupTypeUtil::toLookupType)
+                    .collect(toList());
+
+        });
+        for (usace.cwms.db.dao.ifc.cat.LookupType lookupType : contractTypeList) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: More double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        List<WaterUserType> waterUserTypes = connectionResult(dsl, c -> {
+            setOffice(c, projectLocation.getOfficeId());
+            LOCATION_REF_T projectLocationRef =  LocationTypeUtil.toLocationRefT(locationRefType);
+            WATER_USER_TAB_T waterUserObjTs = CWMS_WATER_SUPPLY_PACKAGE.call_RETRIEVE_WATER_USERS(
+                DSL.using(c).configuration(), projectLocationRef);
+            return waterUserObjTs.stream()
+                .map(WaterUserTypeUtil::toWaterUserType)
+                .collect(toList());
+        });
+        for (WaterUserType waterUserType : waterUserTypes) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        List<WaterUserType> userList = connectionResult(dsl, c -> {
+            setOffice(c, projectLocation.getOfficeId());
+            LOCATION_REF_T projectLocationRef =  LocationTypeUtil.toLocationRefT(locationRefType);
+            WATER_USER_TAB_T waterUserObjTs = CWMS_WATER_SUPPLY_PACKAGE.call_RETRIEVE_WATER_USERS(
+                    DSL.using(c).configuration(), projectLocationRef);
+            return waterUserObjTs.stream()
+                    .map(WaterUserTypeUtil::toWaterUserType)
+                    .collect(toList());
+        });
+        for (WaterUserType waterUser : userList) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            if (waterUser.getEntityName().equals(entityName)) {
+                return WaterSupplyUtils.map(waterUser, projectLocation);
+            }
+        }
+        return null;
+    }
+
+    public void storeWaterContract(WaterUserContract waterContractType, boolean failIfExists, boolean ignoreNulls) {
+        List<WaterUserContractType> waterUserContractTypeModified = new ArrayList<>();
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: why is this called modified?

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+    public void renameWaterContract(WaterUser waterUser, String oldContractName,
+            String newContractName) {
+
+        WaterUserContractRefType waterUserContractRefType =
+                WaterSupplyUtils.map(waterUser, waterUser.getProjectId(), newContractName);
+
+        connection(dsl, c -> {
+            setOffice(c, waterUser.getProjectId().getOfficeId());
+            WATER_USER_OBJ_T waterUserT = WaterUserTypeUtil.toWaterUserT(waterUserContractRefType.getWaterUserType());
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        WaterUser waterUser = new WaterUser.Builder().withEntityName(contract.getWaterUser().getEntityName())
+                .withProjectId(contract.getWaterUser().getProjectId())
+                .withWaterRight(contract.getWaterUser().getWaterRight()).build();
+
+        WaterUserContractRefType waterUserContractRefTypeModified =
+                WaterSupplyUtils.map(waterUser, waterUser.getProjectId(), contract.getContractId().getName());
+
+        connection(dsl, c -> {
+            setOffice(c, contract.getOfficeId());
+            WATER_USER_OBJ_T waterUserT = WaterUserTypeUtil.toWaterUserT(waterUserContractRefTypeModified
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+import cwms.cda.data.dto.watersupply.WaterUserContract;
+import java.time.ZoneId;
+import usace.cwms.db.dao.ifc.loc.LocationRefType;
+import usace.cwms.db.dao.ifc.loc.LocationType;
+import usace.cwms.db.dao.ifc.watersupply.WaterUserContractRefType;
+import usace.cwms.db.dao.ifc.watersupply.WaterUserContractType;
+import usace.cwms.db.dao.ifc.watersupply.WaterUserType;
+
+
+public class WaterSupplyUtils {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: make class final

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+import cwms.cda.data.dto.LookupType;
+import cwms.cda.data.dto.watersupply.PumpType;
+import cwms.cda.data.dto.watersupply.WaterSupplyPump;
+import cwms.cda.data.dto.watersupply.WaterUser;
+import cwms.cda.data.dto.watersupply.WaterUserContract;
+import java.time.ZoneId;
+import usace.cwms.db.dao.ifc.loc.LocationRefType;
+import usace.cwms.db.dao.ifc.loc.LocationType;
+import usace.cwms.db.dao.ifc.watersupply.WaterUserContractRefType;
+import usace.cwms.db.dao.ifc.watersupply.WaterUserContractType;
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: all mappings should be from jOOQ codegen of the database UDT's to/from CDA DTO's. there doesn't need to be any intermediary `cwms-db-jooq`/`cwms-db-dao` classes

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                    .withMapLabel(locationType.getMapLabel())
+                    .withPublishedLatitude(locationType.getPublishedLatitude())
+                    .withPublishedLongitude(locationType.getPublishedLongitude())
+                    .withOfficeId(locationType.getLocationRef().getOfficeId())
+                    .build())
+                .withPumpType(pumpType).build();
+    }
+
+    public static LocationType map(WaterSupplyPump pump) {
+        return new LocationType(new LocationRefType(pump.getPumpLocation().getName(),
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: LocationUtil does the correct type mapping for you already

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+        });
+    }
+
+    private void cleanupContractRoutine(WaterUserContract contract) throws Exception {
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.deleteWaterContract(contract, DELETE_ACTION);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+        cleanupUserRoutine(contract.getWaterUser());
+    }
+
+    private void cleanupUserRoutine(WaterUser user) throws Exception {
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.deleteWaterUser(user.getProjectId(), user.getEntityName(), DELETE_ACTION);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+            dao.storeWaterUser(contract.getWaterUser(), true);
+            dao.storeWaterContract(contract, true, false);
+            dao.removePumpFromContract(contract, contract.getPumpInLocation().getPumpLocation().getName(),
+                    "IN", false);
+            List<WaterUserContract> contracts = dao.getAllWaterContracts(contract.getWaterUser()
+                    .getProjectId(), contract.getWaterUser().getEntityName());
+            WaterUserContract retrievedContract = contracts.get(0);
+            assertNotNull(retrievedContract);
+            assertNull(retrievedContract.getPumpInLocation());
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(contract.getWaterUser(), true);
+            dao.storeWaterContract(contract, true, false);
+            dao.deleteWaterContract(contract, DELETE_ACTION);
+            List<WaterUserContract> contracts = dao.getAllWaterContracts(contract.getWaterUser()
+                    .getProjectId(), contract.getWaterUser().getEntityName());
+            assertTrue(contracts.isEmpty());
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+    void testDeleteWaterUser() throws Exception {
+        WaterUser newUser = buildTestWaterUser("TEST DELETE USER");
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(newUser, true);
+            dao.deleteWaterUser(newUser.getProjectId(), newUser.getEntityName(), DELETE_ACTION);
+            assertNull(dao.getWaterUser(newUser.getProjectId(), newUser.getEntityName()));
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+                    .withProjectId(oldContract.getWaterUser().getProjectId())
+                    .withWaterRight(oldContract.getContractId().getName()).build();
+            dao.renameWaterContract(user, oldContract.getContractId().getName(),
+                    renamedContract.getContractId().getName());
+            List<WaterUserContract> retrievedContracts = dao.getAllWaterContracts(
+                    renamedContract.getWaterUser().getProjectId(), renamedContract.getWaterUser()
+                            .getEntityName());
+            assertFalse(retrievedContracts.isEmpty());
+            DTOMatch.assertMatch(retrievedContracts.get(0), renamedContract);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+        WaterUser updatedUser = buildTestWaterUser("TEST RENAME USER 2");
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(newUser, true);
+            dao.renameWaterUser(newUser.getEntityName(), updatedUser.getEntityName(), newUser.getProjectId());
+            WaterUser retrievedUser = dao.getWaterUser(newUser.getProjectId(), updatedUser.getEntityName());
+            DTOMatch.assertMatch(updatedUser, retrievedUser);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+            contractType.add(new LookupType.Builder()
+                    .withTooltip("Test Tooltip 2")
+                    .withActive(true)
+                    .withDisplayValue("Test Display Value 2")
+                    .withOfficeId(OFFICE_ID).build());
+            dao.storeWaterContractTypes(contractType, false);
+            List<LookupType> results = dao.getAllWaterContractTypes(OFFICE_ID);
+            DTOMatch.assertMatch(contractType.get(0), results.get(0));
+            DTOMatch.assertMatch(contractType.get(1), results.get(1));
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(contract.getWaterUser(), true);
+            dao.storeWaterContract(contract, true, true);
+            List<WaterUserContract> retrievedContract = dao.getAllWaterContracts(contract.getWaterUser().getProjectId(), contract.getWaterUser().getEntityName());
+            WaterUserContract result = retrievedContract.get(0);
+            DTOMatch.assertMatch(contract, result);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+        WaterUser newUser = buildTestWaterUser(entityName);
+
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(newUser, true);
+            WaterUser retrievedUser = dao.getWaterUser(projectLocation, entityName);
+            DTOMatch.assertMatch(newUser, retrievedUser);
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(buildTestWaterUser("TEST RETRIEVE USER 1"), false);
+            dao.storeWaterUser(buildTestWaterUser("TEST RETRIEVE USER 2"), false);
+            List<WaterUser> results = dao.getAllWaterUsers(new CwmsId.Builder().withOfficeId(OFFICE_ID).withName(testLocation.getName()).build());
+            DTOMatch.assertMatch(results.get(0), buildTestWaterUser("TEST RETRIEVE USER 1"));
+            DTOMatch.assertMatch(results.get(1), buildTestWaterUser("TEST RETRIEVE USER 2"));
+        });
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: add the correct web user

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    @AfterAll
+    static void tearDown() throws Exception {
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            ProjectDao projectDao = new ProjectDao(ctx);
+            projectDao.delete(OFFICE_ID, testProject.getLocation().getName(), DeleteRule.DELETE_ALL);
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java:104]: This cleans up all related water supply data and the location?

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+    @Test
+    void testDeleteWaterUser() throws Exception {
+        WaterUser newUser = buildTestWaterUser("TEST DELETE USER");
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(newUser, true);
+            dao.deleteWaterUser(newUser.getProjectId(), newUser.getEntityName(), DELETE_ACTION);
+            assertNull(dao.getWaterUser(newUser.getProjectId(), newUser.getEntityName()));
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: shouldn't this throw a not found exception?

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+            dao.storeWaterContract(oldContract, true, false);
+            WaterUser user = new WaterUser.Builder().withEntityName(oldContract.getWaterUser().getEntityName())
+                    .withProjectId(oldContract.getWaterUser().getProjectId())
+                    .withWaterRight(oldContract.getContractId().getName()).build();
+            dao.renameWaterContract(user, oldContract.getContractId().getName(),
+                    renamedContract.getContractId().getName());
+            List<WaterUserContract> retrievedContracts = dao.getAllWaterContracts(
+                    renamedContract.getWaterUser().getProjectId(), renamedContract.getWaterUser()
+                            .getEntityName());
+            assertFalse(retrievedContracts.isEmpty());
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: should also test that the old contract is gone

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+        WaterUser newUser = buildTestWaterUser("TEST RENAME USER");
+        WaterUser updatedUser = buildTestWaterUser("TEST RENAME USER 2");
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(newUser, true);
+            dao.renameWaterUser(newUser.getEntityName(), updatedUser.getEntityName(), newUser.getProjectId());
+            WaterUser retrievedUser = dao.getWaterUser(newUser.getProjectId(), updatedUser.getEntityName());
+            DTOMatch.assertMatch(updatedUser, retrievedUser);
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java:255]: check to make sure the old water user is gone

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 28]
```diff
@@ -0,0 +1,293 @@
...
+import usace.cwms.db.jooq.dao.util.WaterUserTypeUtil;
+
+
+
+public class WaterContractDao extends JooqDao<WaterUserContract> {
+    public WaterContractDao(DSLContext dsl) {
+        super(dsl);
+    }
+
+    public List<WaterUserContract> getAllWaterContracts(CwmsId projectLocation, String entityName) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java:57]: no use case for retrieving a single water user contract?

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 1 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: Removed USACE DAO references

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 2 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: Removed double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 3 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: Altered mapping as suggested

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 4 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: Removed double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 7 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java]: Removed as part of double mapping cleanup

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java (hunk 10 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: Changed to final

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java (hunk 11 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: Removed double mapping

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java (hunk 12 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java]: Removed duplicate method

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java (hunk 13 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: Added for all DB calls

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java (hunk 24 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java:104]: Added location cleanup, confirmed contract and water user cleanup are run at end of each test

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java (hunk 25 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: Added not found exception throws to getWaterUser and getWaterContract

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java (hunk 26 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: Added check for contract no longer in DB

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java (hunk 27 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java:255]: Added check for water user no longer in DB

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 28 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java:57]: Added single water contract retrieval with test coverage

File: cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java [hunk 29]
```diff
@@ -0,0 +1,422 @@
...
+        db.connection(c -> {
+            DSLContext ctx = getDslContext(c, OFFICE_ID);
+            WaterContractDao dao = new WaterContractDao(ctx);
+            dao.storeWaterUser(buildTestWaterUser("TEST RETRIEVE USER 1"), false);
+            dao.storeWaterUser(buildTestWaterUser("TEST RETRIEVE USER 2"), false);
+            List<WaterUser> results = dao.getAllWaterUsers(new CwmsId.Builder().withOfficeId(OFFICE_ID).withName(testLocation.getName()).build());
+            DTOMatch.assertMatch(results.get(0), buildTestWaterUser("TEST RETRIEVE USER 1"));
+            DTOMatch.assertMatch(results.get(1), buildTestWaterUser("TEST RETRIEVE USER 2"));
+        }, CwmsDataApiSetupCallback.getWebUser());
+        cleanupUserRoutine(buildTestWaterUser("TEST RETRIEVE USER 1"));
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: cleanup will not get called if the test fails, recommend adding to the @AfterEach or @AfterAll

Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/watersupply/WaterContractDaoTestIT.java]: Moved cleanup into AfterEach function

**Comment by adamkorynta** [comment]
The one integration test failure looks like a flaky stream test unrelated to this PR. Will look into separately.

**Review by adamkorynta — APPROVED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+import usace.cwms.db.jooq.codegen.udt.records.LOOKUP_TYPE_OBJ_T;
+import usace.cwms.db.jooq.codegen.udt.records.LOOKUP_TYPE_TAB_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_CONTRACT_OBJ_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_CONTRACT_REF_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_CONTRACT_TAB_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_OBJ_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_TAB_T;
+
+
+public final class WaterSupplyUtils {
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java:49]: Cant this whole class be package scoped?

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+import usace.cwms.db.jooq.codegen.packages.CWMS_WATER_SUPPLY_PACKAGE;
+import usace.cwms.db.jooq.codegen.udt.records.LOCATION_REF_T;
+import usace.cwms.db.jooq.codegen.udt.records.LOOKUP_TYPE_TAB_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_CONTRACT_REF_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_CONTRACT_TAB_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_OBJ_T;
+import usace.cwms.db.jooq.codegen.udt.records.WATER_USER_TAB_T;
+
+
+public class WaterContractDao extends JooqDao<WaterUserContract> {
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java:52]: Shouldn't this class should be final? Unless you plan on supporting subclassing?

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java (hunk 30 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterSupplyUtils.java:49]: Yes, will update that in the controller PR

File: cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java (hunk 31 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/watersupply/WaterContractDao.java:52]: Yes, it should be. Will change to final in controller PR

**Pull request merged**